bittide-instances/exe/post-vex-riscv-test/Main.hs

```diff
@@ -42,81 +43,97 @@ GDB: GNU Debugger. This program will connect to the OpenOCD server and is able
-      hSetBuffering openOcdStdErr LineBuffering
-      expectLine openOcdStdErr waitForHalt
-
-      -- XXX: Picocom doesn't immediately clean up after closing, because it
-      --      spawns as a child of the shell (start.sh). We could use 'exec' to
-      --      make sure the intermediate shell doesn't exist, but this causes
-      --      the whole test program to exit with signal 15 (??????).
-      withCreateProcess picocomProc $ \maybePicocomStdIn maybePicocomStdOut maybePicocomStdErr _ -> do
-        let
-          picocomStdIn = fromJust maybePicocomStdIn
-          picocomStdOut = fromJust maybePicocomStdOut
-
-          -- Create function to log the output of the processes
-          loggingSequence = do
-            threadDelay 1_000_000 -- Wait 1 second for data loggers to catch up
-            putStrLn "Picocom stdout"
-            picocomOut <- readRemainingChars picocomStdOut
-            putStrLn picocomOut
-            case maybePicocomStdErr of
-              Nothing -> pure ()
-              Just h -> do
-                putStrLn "Picocom StdErr"
-                readRemainingChars h >>= putStrLn
-
-          tryWithTimeout :: String -> Int -> IO a -> IO a
-          tryWithTimeout actionName dur action = do
-            result <- timeout dur action
-            case result of
-              Nothing -> do
-                loggingSequence
-                assertFailure $ "Timeout while performing action: " <> actionName
-              Just r -> pure r
-
-        hSetBuffering picocomStdIn LineBuffering
-        hSetBuffering picocomStdOut LineBuffering
-
-        tryWithTimeout "Waiting for \"Terminal ready\"" 10_000_000 $
-          waitForLine picocomStdOut "Terminal ready"
-
-        withCreateProcess gdbProc $ \_ (fromJust -> gdbStdOut) _ _ -> do
-          -- Wait for GDB to program the FPGA. If successful, we should see
-          -- "going in echo mode" in the picocom output.
-          hSetBuffering gdbStdOut LineBuffering
-          tryWithTimeout "Waiting for \"Going in echo mode!\"" 10_000_000 $
-            waitForLine picocomStdOut "Going in echo mode!"
-
-          -- Wait for GDB to reach its last command - where it will wait indefinitely
-          tryWithTimeout "Waiting for \"> continue\"" 10_000_000 $
-            waitForLine gdbStdOut "> continue"
-
-          -- Test UART echo
-          hPutStrLn picocomStdIn "Hello, UART!"
-          tryWithTimeout "Waiting for \"Hello, UART!\"" 10_000_000 $
-            waitForLine picocomStdOut "Hello, UART!"
+  results <- mapM (\d -> catch (testGdb d) catchHUnitFailure) demoRigInfo
+  print results
+  assertBool "Not all FPGAs were correctly programmed" (and results)
+ where
+  -- We use this solution so the test stops when it encounters a problem
+  catchHUnitFailure :: HUnitFailure -> IO Bool
+  catchHUnitFailure (HUnitFailure _ msg) = do
+    putStrLn $ "Test failed with: " <> msg
+    pure False
```

Comment:
Perhaps a good idea to also pass `d` to `catchHUnitFailure` and print some more info regarding the device for which the test failed.